_posts/2017-11-27-Babel-7-and-TypeScript.markdown

```diff
@@ -0,0 +1,261 @@
+```
+```typescript
+// app/server.ts
+console.log('hi!')
+```
+
+Running
+
+```sh
+yarn start
+$ babel-node index.js
+
+hi!
+✨  Done in 1.88s.
+```
+
+Everything seems to be working; our `.js` entrypoint is configured to support `.ts` extensions and we kick off the boot process.
+
+Let's now try to import a file from within `app/server.ts`:
+
+```typescript
+import path from 'path'
+console.log(`Hello ${path.resolve(process.cwd())}!`)
+```
+
+```sh
+yarn start
+$ yarn run v1.3.2
+$ babel-node index.js
+sites/src/index.tsx:1
+(function (exports, require, module, __filename, __dirname) { import path from 'path'
+                                                              ^^^^^^
+
+SyntaxError: Unexpected token import
+```
+
+Maybe my `tsconfig.json` file is misconfigured?
+
+```json
+{
+  "compilerOptions": {
+    "module": "es2015"
+  }
+}
+```
+
+Nope, all good. How about my `.babelrc`?
+
+```json
+{
+  "presets": [
+    ["@babel/env", {
+      "targets": {
+        "browsers": ["last 2 versions"]
+      }
+    }],
+    "@babel/stage-3",
+    "@babel/react",
+    "@babel/typescript"
+  ]
+}
+```
+
+We're using [`@babel/preset-env`](https://github.com/babel/babel/tree/master/packages/babel-preset-env) which handles selecting the JS features we need, so thats not it. And anyways, doesn't TypeScript support `ES2015` modules right out of the box?
+
+Continuing, how about specifying the extension list directly in `package.json`:
+
+```json
+"start": "babel-node --extensions '.ts,.tsx' index.js"
+```
+
+Still no go :(
+
+Last try: Create a new entrypoint file that uses a `.ts` extension and then use _that_ to boot the rest of the app:
+
+```json
+"start": "babel-node --extensions '.ts,.tsx' index.ts"
+```
+```typescript
+// index.ts
+import './app/server'
+```
+```sh
+yarn start
+$ yarn run v1.3.2
+$ babel-node index.js
+Hello /sites!
+```
+
+Once this change was in place, we could ditch `@babel/register` and instead rely on the `--extensions` configuration from `package.json`.
+
+**NOTE:** If you're using [`babel-plugin-module-resolver`](https://github.com/tleunen/babel-plugin-module-resolver) to support absolute path imports make sure to update the `extensions` [option](https://github.com/tleunen/babel-plugin-module-resolver#options) with `.ts` and `.tsx`.
+
+### 3) Type-Checking
+
+Lastly, since Babel 7 is now responsible for compiling our TypeScript files we no longer need to rely on TypeScript's own `tsc` compiler to output JavaScript and instead just use it to type-check our code. Again, in `package.json`:
+
+```
+"typecheck": "tsc"
+```
+
+This reads in settings located in `tsconfig.json`:
+```json
+{
+  "compilerOptions": {
+    "allowJs": true,
+    "allowSyntheticDefaultImports": true,
+    "baseUrl": "./src",
+    "jsx": "react",
+    "module": "es2015",
+    "noEmit": true,
+    "noUnusedLocals": true,
+    "noUnusedParameters": true,
+    "pretty": true,
+    "rootDirs": [
+      "./src"
+    ],
+    "sourceMap": true
+  }
+}
+```
```

Comment:
Seeing as you link to the full hello world project, I think it would be fine to omit all unrelated entries for sake of clarity and leave just the `emit` _and_ `pretty` options, it’s for instance not clear that the latter is related to having nice type-checker output.